plat/common/arm/lcpu.c

```diff
@@ -0,0 +1,296 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <stdint.h>
+#include <uk/plat/lcpu.h>
+#include <uk/plat/io.h>
+#include <arm/irq.h>
+#include <arm/smccc.h>
+#include <arm/arm64/cpu.h>
+#include <gic/gic.h>
+#include <libfdt.h>
+#include <ofw/fdt.h>
+#include <stdint.h>
+
+/**
+ * The number of cells for the size field should be 0 in cpu nodes.
+ * The number of cells in the address field is set by default to 2 in cpu
+ * nodes. This is also the maximum number of cells for the address field that
+ * we can currently support.
+ **/
+#define FDT_SIZE_CELLS_DEFAULT 0
+#define FDT_ADDR_CELLS_DEFAULT 2
+#define CPU_ID_MASK 0xff00ffffffUL
+
+static void *dtb;
+void lcpu_start(struct lcpu *cpu);
+static __uptr lcpu_start_addr;
+extern struct _gic_dev *gic;
+
+void ukplat_lcpu_enable_irq(void)
+{
+	local_irq_enable();
+}
+
+void ukplat_lcpu_disable_irq(void)
+{
+	local_irq_disable();
+}
+
+void ukplat_lcpu_halt_irq(void)
+{
+	ukplat_lcpu_enable_irq();
+	halt();
+	ukplat_lcpu_disable_irq();
+}
+
+unsigned long ukplat_lcpu_save_irqf(void)
+{
+	unsigned long flags;
+
+	local_irq_save(flags);
+
+	return flags;
+}
+
+void ukplat_lcpu_restore_irqf(unsigned long flags)
+{
+	local_irq_restore(flags);
+}
+
+int ukplat_lcpu_irqs_disabled(void)
+{
+	return irqs_disabled();
+}
+
+void ukplat_lcpu_irqs_handle_pending(void)
+{
+	// TODO
+}
+
+__lcpuid lcpu_arch_id(void)
+{
+	__u64 mpidr_reg;
+
+	mpidr_reg = SYSREG_READ64(mpidr_el1);
+
+	/* return the affinity bits for the current core */
+	return mpidr_reg & CPU_ID_MASK;
+}
+
+#ifdef CONFIG_HAVE_SMP
+int lcpu_arch_init(struct lcpu *this_lcpu)
+{
+	int ret = 0;
+
+	/* Initialize the interrupt controller for non-bsp cores */
+	if (!lcpu_is_bsp(this_lcpu)) {
+		ret = _dtb_init_gic(dtb, &gic);
+		UK_ASSERT(ret == 0);
+		ret = gic->ops.initialize();
+		UK_ASSERT(ret == 0);
```

Comment:
Hmm I think you are right, the `init_gic()` calls on APs are useless. Theoretically, race conditions should be avoided because the `gic-v2.c` and `gic-v3.c` sources document the probing functions by saying: `First time must not be called from multiple CPUs in parallel`. Therefore, the platform code should make sure that there should be only one core initializing the cpu at the beginning, then any other parallel accesses are race free and only check that the gic was successfully initialized.

However, this is just to much for what we need (halting secondary cores is GIC initialization failed on the bsp). All we need here is checking `gic != NULL`, I don't know why I haven't done this in the first place.

Reply:
As we should not reach SMP initialization if GIC initialization on the BSP fails, it is probably sufficient to just have a `UK_ASSERT(gic);` before accessing it.

Reply:
Right, I would add it at the beginning of `lcpu_mp_init()`, sounds good?

Reply:
I suggest to have always a tendency to put ASSERTS where the actual condition must be true. So while `gic` should already be initialized in `arch_mp_init`, I would place it just before the `gic->`. 

Another example is something like this:
```C
void myfunc(something_t *obj1_not_null, something_t *obj2_not_null)
{
    UK_ASSERT(obj1_not_null != NULL);
    obj1_not_null->func(obj2_not_null);
}

void obj_func(something_t *arg) {
    UK_ASSERT(arg != NULL);
    arg->X ...
}
```
Although obj2 should not be NULL, too, it is not dereferenced. But in any case `obj_func` is responsible for making sure that its parameters meet its constraints so it performs the check itself. `myfunc` would only check if it dereferences or otherwise uses obj2 in a situation where it must guarantee a non-NULL value.

Reply:
> it is probably sufficient to just have a UK_ASSERT(gic); before accessing it.

@marcrittinghaus, actually `UK_ASSERT()` is more sufficient for logic errors right? And it will not be there without ukdebug. While GIC initialization depends on hardware and may fail for arbitrary reasons, so perhaps an if-based check is more appropriate here?

Reply:
Ok, that is a valid argument. I also see `if` for runtime errors and `UK_ASSERT` for things that violate developer expectations and should be regarded a bug.

In that sense, I was assuming that we would not get this far into the boot (i.e., access the GIC on a secondary core) if we did not do proper GIC initialization on the BSP first

Reply:
But I have no strong preference. An `if` is fine too

Reply:
>  Hmm I think you are right, the init_gic() calls on APs are useless. Theoretically, race conditions should be avoided because the gic-v2.c and gic-v3.c sources document the probing functions by saying: First time must not be called from multiple CPUs in parallel. Therefore, the platform code should make sure that there should be only one core initializing the cpu at the beginning, then any other parallel accesses are race free and only check that the gic was successfully initialized.

Without looking at the gic sources, IIRC it's the Distributor that is initialized once, but the CPU Interface must be initialized per core, so most likely you need to call `init_gic()` from all cores.

PS: I have never come across a SoC that uses different GIC versions. Usually what you have an auxiliary Cortex-R, in which case you would probably treat it separately than the main Cortex-A cluster (won't be part of your SMP setup) but it will be running some RTOS to handle real-time payloads.